Datasource Connection issues while migrating database from 17.0.1 to 21.0.0

### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

account/api

### Describe the bug

I have a Keycloak (wildfly) instance with version 17.0.1 having ~600 realms and while migrating the same to keycloak-21.0.1, keycloak instance fails to start with below exception :

2023-03-30 23:22:37,405 DEBUG [org.hibernate.SQL] (main) select cliententi0_.ID as col_0_0_ from CLIENT cliententi0_ where cliententi0_.CLIENT_ID=? and cliententi0_.REALM_ID=?
2023-03-30 23:22:37,406 DEBUG [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (main) could not prepare statement [select cliententi0_.ID as col_0_0_ from CLIENT cliententi0_ where cliententi0_.CLIENT_ID=? and cliententi0_.REALM_ID=?]: java.sql.SQLException: Connection is closed
        at io.agroal.pool.wrapper.ConnectionWrapper.lambda$static$0(ConnectionWrapper.java:52)
        at com.sun.proxy.$Proxy62.prepareStatement(Unknown Source)
        at io.agroal.pool.wrapper.ConnectionWrapper.prepareStatement(ConnectionWrapper.java:659)
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$5.doPrepare(StatementPreparerImpl.java:149)
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:176)
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.prepareQueryStatement(StatementPreparerImpl.java:151)
        at org.hibernate.loader.Loader.prepareQueryStatement(Loader.java:2122)
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2059)
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2037)
        at org.hibernate.loader.Loader.doQuery(Loader.java:956)
        at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:357)
        at org.hibernate.loader.Loader.doList(Loader.java:2868)
        at org.hibernate.loader.Loader.doList(Loader.java:2850)
        at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2682)
        at org.hibernate.loader.Loader.list(Loader.java:2677)
        at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:540)
        at org.hibernate.hql.internal.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:400)
        at org.hibernate.engine.query.spi.HQLQueryPlan.performList(HQLQueryPlan.java:219)
        at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1459)
        at org.hibernate.query.internal.AbstractProducedQuery.doList(AbstractProducedQuery.java:1649)
        at org.hibernate.query.internal.AbstractProducedQuery.list(AbstractProducedQuery.java:1617)
        at org.hibernate.query.Query.getResultList(Query.java:165)
        at org.keycloak.models.jpa.JpaRealmProvider.getClientByClientId(JpaRealmProvider.java:760)
        at org.keycloak.storage.ClientStorageManager.getClientByClientId(ClientStorageManager.java:149)
        at org.keycloak.models.cache.infinispan.RealmCacheSession.getClientByClientId(RealmCacheSession.java:1209)
        at org.keycloak.models.cache.infinispan.RealmAdapter.getClientByClientId(RealmAdapter.java:815)
        at org.keycloak.models.utils.KeycloakModelUtils.lambda$getClientScopeByName$13(KeycloakModelUtils.java:819)
        at java.base/java.util.Optional.orElseGet(Optional.java:369)
        at org.keycloak.models.utils.KeycloakModelUtils.getClientScopeByName(KeycloakModelUtils.java:819)
        at org.keycloak.protocol.oidc.OIDCLoginProtocolFactory.addAcrClientScope(OIDCLoginProtocolFactory.java:372)
        at org.keycloak.services.migration.DefaultMigrationProvider.addOIDCAcrClientScope(DefaultMigrationProvider.java:107)
        at org.keycloak.migration.migrators.MigrateTo18_0_0.migrateRealm(MigrateTo18_0_0.java:60)
        at org.keycloak.migration.migrators.MigrateTo18_0_0.lambda$migrate$0(MigrateTo18_0_0.java:46)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
        at org.hibernate.query.spi.StreamDecorator.forEach(StreamDecorator.java:153)
        at org.keycloak.utils.ClosingStream.forEach(ClosingStream.java:128)
        at org.keycloak.migration.migrators.MigrateTo18_0_0.migrate(MigrateTo18_0_0.java:46)
        at org.keycloak.storage.datastore.LegacyMigrationManager.migrate(LegacyMigrationManager.java:135)
        at org.keycloak.migration.MigrationModelManager.migrate(MigrationModelManager.java:33)
        at 
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:93)
        at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:88)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
        at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)

2023-03-30 23:22:37,407 WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (main) SQL Error: 0, SQLState: null
2023-03-30 23:22:37,407 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (main) Connection is closed
2023-03-30 23:22:37,409 DEBUG [org.hibernate.resource.jdbc.internal.ResourceRegistryStandardImpl] (main) Exception clearing maxRows/queryTimeout [Statement is closed]
2023-03-30 23:22:37,409 DEBUG [org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (main) Initiating JDBC connection release from afterTransaction
2023-03-30 23:22:37,409 DEBUG [org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl] (main) Initiating JDBC connection release from afterTransaction
2023-03-30 23:22:37,424 DEBUG [org.hibernate.engine.transaction.internal.TransactionImpl] (main) On TransactionImpl creation, JpaCompliance#isJpaTransactionComplianceEnabled == false
2023-03-30 23:22:37,425 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (main) JtaTransactionWrapper rollback
2023-03-30 23:22:37,426 WARN  [com.arjuna.ats.arjuna] (main) ARJUNA012077: Abort called on already aborted atomic action 0:ffff0a3c1e68:a4bd:6425cb36:0
2023-03-30 23:22:37,426 DEBUG [org.keycloak.transaction.JtaTransactionWrapper] (main) JtaTransactionWrapper end
2023-03-30 23:22:37,460 DEBUG [org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl] (main) HHH000420: Closing un-released batch
2023-03-30 23:22:37,462 DEBUG [org.infinispan.manager.DefaultCacheManager] (main) Stopping cache manager XXXXXX-39174
2023-03-30 23:22:37,463 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache authenticationSessions on XXXXX-39174
2023-03-30 23:22:37,463 DEBUG [org.infinispan.topology.LocalTopologyManagerImpl] (main) Node XXXXX-39174 leaving cache authenticationSessions
2023-03-30 23:22:37,464 DEBUG [org.infinispan.partitionhandling.impl.PreferAvailabilityStrategy] (main) The last node of cache authenticationSessions left
2023-03-30 23:22:37,468 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache realms on keycloak
2023-03-30 23:22:37,474 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache sessions on XXXXX-39174
2023-03-30 23:22:37,474 DEBUG [org.infinispan.topology.LocalTopologyManagerImpl] (main) Node XXXXX-39174 leaving cache sessions
2023-03-30 23:22:37,474 DEBUG [org.infinispan.partitionhandling.impl.PreferAvailabilityStrategy] (main) The last node of cache sessions left
2023-03-30 23:22:37,475 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache authorizationRevisions on keycloak
2023-03-30 23:22:37,476 DEBUG [org.infinispan.transaction.impl.TransactionTable] (main) Wait for on-going transactions to finish for 30 seconds.
2023-03-30 23:22:37,476 DEBUG [org.infinispan.transaction.impl.TransactionTable] (main) All transactions terminated
2023-03-30 23:22:37,476 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache clientSessions on XXXXX-39174
2023-03-30 23:22:37,476 DEBUG [org.infinispan.topology.LocalTopologyManagerImpl] (main) Node xXXXX-39174 leaving cache clientSessions
2023-03-30 23:22:37,476 DEBUG [org.infinispan.partitionhandling.impl.PreferAvailabilityStrategy] (main) The last node of cache clientSessions left
2023-03-30 23:22:37,478 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache keys on keycloak
2023-03-30 23:22:37,478 DEBUG [org.infinispan.cache.impl.CacheImpl] (main) Stopping cache work on XXXXX-39174
202
org.keycloak.quarkus.runtime.storage.legacy.database.LegacyJpaConnectionProviderFactory.migrateModel(LegacyJpaConnectionProviderFactory.java:216)
2023-03-30 23:22:37,565 DEBUG [io.quarkus.runtime.Application] (main) Stopping application
2023-03-30 23:22:37,566 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
2023-03-30 23:22:37,566 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: org.hibernate.HibernateException: Transaction was rolled back in a different thread!
2023-03-30 23:22:37,566 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Transaction was rolled back in a different thread!
2 

### Version

21.0.0

### Expected behavior

Keycloak 17.0.1 database should be migrated to keycloak 21.0.0 and keycloak instance should started 

### Actual behavior

Migration of Keycloak 17.0.1 (Wildfly) to Keycloak-21.0.0 (quarkus) on database with ~600 realms is failing to migrate and keycloak server fails to start.

### How to Reproduce?

1. Should have an Keycloak instance (17.0.1) with ~500 realms 
2. Install Keycloak 21.0.0 and point to Keycloak-17.0.1 database
3. Start server

### Anything else?

_No response_

very strage. i too see similar thing when upgrading to keyclok 21.0.2

```
        at org.hibernate.query.Query.getResultStream(Query.java:1140)
        at org.keycloak.models.jpa.JpaRealmProvider.getClientScopes(JpaRealmProvider.java:987)
        at org.keycloak.storage.ClientStorageManager.getClientScopes(ClientStorageManager.java:212)
.......
.....
.....

Caused by: java.sql.SQLException: Connection is closed
        at io.agroal.pool.wrapper.ConnectionWrapper.lambda$static$0(ConnectionWrapper.java:52)

```

--- a/docs/documentation/upgrading/topics/changes/changes-26_6_0.adoc
+++ b/docs/documentation/upgrading/topics/changes/changes-26_6_0.adoc
@@ -25,6 +25,16 @@ Configuring + defers validation to the client's standard 'Valid redirect URIs'. 
 
 Administrators must ensure that all configured URIs for the fields listed above use https. Clients attempting to update or register with http in these fields will fail validation when using the 'secure-client-uris' executor.
 
+=== New CLI option overrides Quarkus transaction timeout property
+
+A new CLI option `transaction-default-timeout` has been introduced to configure the default transaction timeout in {project_name}.
+This new option will override any value set in the `quarkus.transaction-manager.default-transaction-timeout` Quarkus property.
+
+While using Quarkus properties directly is not the supported way to configure {project_name}, some users may have used `quarkus.transaction-manager.default-transaction-timeout` to customize the transaction timeout.
+If you are using this Quarkus property, the new `transaction-default-timeout` CLI option will take precedence and override your configuration when upgrading to this version.
+
+To maintain your desired transaction timeout configuration, migrate to using the `transaction-default-timeout` CLI option and remove the `quarkus.transaction-manager.default-transaction-timeout` property from your configuration.
+
 
 // ------------------------ Notable changes ------------------------ //
 == Notable changes
@@ -148,6 +158,16 @@ Authentication failures resulting in an invalid_grant now correctly return an HT
 
 With the `clusterless` feature enabled and `persistent-user-sessions` disabled, a re-index is required due to schema changes.
 {project_name} triggers the re-index on startup, but this may delay server startup depending on the number of sessions stored, and/or return incorrect results during the re-index process.
+
+=== New CLI options for transaction timeout configuration
+
+Two new CLI options are now available to control transaction timeout behavior:
+
+* `transaction-default-timeout`: Controls the transaction timeout for regular transactions.
+The default value is 5 minutes, which is the same value used by the current {project_name} version.
+
+* `transaction-migration-timeout`: Controls the timeout for transactions responsible for migrating the database schema to a new version.
+The default value is 30 minutes, as schema migrations may take a while to complete for large tables.
 
 // ------------------------ Deprecated features ------------------------ //
 == Deprecated features

--- a/docs/documentation/upgrading/topics/migrate_db.adoc
+++ b/docs/documentation/upgrading/topics/migrate_db.adoc
@@ -2,7 +2,7 @@
 
 == Migrating the database
 
-{project_name} can automatically migrate the database schema, or you can choose to do it manually. By default the
+{project_name} can automatically migrate the database schema, or you can choose to do it manually. By default, the
 database is automatically migrated when you start the new installation for the first time.
 
 [NOTE]
@@ -14,6 +14,18 @@ Before you migrate the database, shut down all {project_name} nodes running the 
 ====
 Migration is not supported with the default H2 based `dev-file` database type.
 ====
+
+=== Migration timeout configuration
+
+The migration time of the database schema and its contents can vary depending on hardware, release and the amount of data.
+By default {project_name} sets a migration timeout of 30 minutes.
+
+To change the migration timeout, use the option `transaction-migration-timeout` to configure a longer value if your environment requires this.
+
+[source,bash]
+----
+kc.[sh|bat] start --transaction-migration-timeout=60m
+----
 
 === Automatic relational database migration
 
@@ -55,6 +67,9 @@ default `connections-jpa` provider:
 kc.[sh|bat] start --spi-connections-jpa--quarkus--migration-export=<path>/<file.sql>
 ----
 
-For further details on how to apply this file to the database, see the documentation for your relational database.
 After the changes have been written to the file, the server exits.
 
+For further details on how to run the statements in the file against the database, see the documentation for your relational database.
+
+Once you have run the statements, you can start the new {project_name} version.
+On the first startup, it might perform additional data migrations that are required in addition to the schema migrations.

--- a/docs/guides/server/db.adoc
+++ b/docs/guides/server/db.adoc
@@ -412,6 +412,35 @@ XA recovery defaults to enabled and will use the file system location `KEYCLOAK_
 
 NOTE: Enabling XA transactions in a containerized environment does not fully support XA recovery unless stable storage is available at that path.
 
+== Configuring transaction timeouts
+
+{project_name} provides two CLI options to control transaction timeout behavior:
+
+`transaction-default-timeout`::
+Controls the transaction timeout for regular transactions.
+The default value is 5 minutes.
++
+To configure a different timeout, for example 10 minutes:
++
+<@kc.start parameters="--transaction-default-timeout=10m"/>
+
+`transaction-migration-timeout`::
+Controls the timeout for transactions responsible for migrating the database schema to a new version.
+The default value is 30 minutes, as schema migrations may take a while to complete for large tables.
++
+To configure a different timeout, for example 1 hour:
++
+<@kc.start parameters="--transaction-migration-timeout=1h"/>
+
+Both options accept values as an ISO 8601 duration, an integer number of seconds, or an integer followed by one of `ms` (milliseconds), `s` (seconds), `m` (minutes), `h` (hours), or `d` (days).
+These options can be configured via CLI, environment variables, or the `conf/keycloak.conf` configuration file.
+
+[NOTE]
+====
+The `transaction-default-timeout` option takes precedence over the unsupported `quarkus.transaction-manager.default-transaction-timeout` Quarkus property.
+If you are using the Quarkus property, migrate to the supported `transaction-default-timeout` option and remove the Quarkus property from your configuration.
+====
+
 == Setting JPA provider configuration option for migrationStrategy
 
 To setup the JPA migrationStrategy (manual/update/validate) you should setup JPA provider as follows:
@@ -557,7 +586,7 @@ quarkus.datasource.user-store.jdbc.transactions=xa
 WARNING: Use Quarkus properties **without quotation** for the datasource name, as properties with the quoted datasource name clash with the new datasource options mapping.
 Therefore, use `quarkus.datasource.user-store.db-kind=h2`, instead of `quarkus.datasource."user-store".db-kind=h2` to prevent any issues.
 
-<@opts.printRelevantOptions includedOptions="db db-* transaction-xa-enabled" excludedOptions="db-*-<datasource>">
+<@opts.printRelevantOptions includedOptions="db db-* transaction-xa-enabled transaction-default-timeout transaction-migration-timeout" excludedOptions="db-*-<datasource>">
 
 === Additional datasources options
 <@opts.includeOptions includedOptions="*-<datasource>"/>

--- a/quarkus/config-api/src/main/java/org/keycloak/config/TransactionOptions.java
+++ b/quarkus/config-api/src/main/java/org/keycloak/config/TransactionOptions.java
@@ -4,6 +4,8 @@ import static org.keycloak.config.WildcardOptionsUtil.getWildcardNamedKey;
 
 public class TransactionOptions {
 
+    public static final String MIGRATION_TRANSACTION_TIMEOUT = "30m";
+
     public static final Option<Boolean> TRANSACTION_XA_ENABLED_DATASOURCE = new OptionBuilder<>("transaction-xa-enabled-<datasource>", Boolean.class)
             .category(OptionCategory.TRANSACTION)
             .description("If set to true, XA for <datasource> datasource will be used.")
@@ -17,6 +19,20 @@ public class TransactionOptions {
             .buildTime(true)
             .defaultValue(Boolean.FALSE)
             .wildcardKey(TRANSACTION_XA_ENABLED_DATASOURCE.getKey())
+            .build();
+
+    public static final Option<String> TRANSACTION_DEFAULT_TIMEOUT = new OptionBuilder<>("transaction-default-timeout", String.class)
+            .category(OptionCategory.TRANSACTION)
+            .description("The default transaction timeout. " + OptionsUtil.DURATION_DESCRIPTION)
+            .buildTime(false)
+            .defaultValue("5m")
+            .build();
+
+    public static final Option<String> TRANSACTION_MIGRATION_TIMEOUT = new OptionBuilder<>("transaction-migration-timeout", String.class)
+            .category(OptionCategory.TRANSACTION)
+            .description("The transaction timeout for database migration transaction. " + OptionsUtil.DURATION_DESCRIPTION)
+            .buildTime(false)
+            .defaultValue(MIGRATION_TRANSACTION_TIMEOUT)
             .build();
 
     public static String getNamedTxXADatasource(String namedProperty) {

--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/TransactionPropertyMappers.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/TransactionPropertyMappers.java
@@ -1,9 +1,14 @@
 package org.keycloak.quarkus.runtime.configuration.mappers;
 
+import java.time.Duration;
 import java.util.List;
 
+import org.keycloak.config.Option;
+import org.keycloak.config.OptionsUtil;
 import org.keycloak.config.TransactionOptions;
+import org.keycloak.quarkus.runtime.cli.PropertyException;
 
+import io.quarkus.runtime.configuration.DurationConverter;
 import io.smallrye.config.ConfigSourceInterceptorContext;
 
 import static org.keycloak.quarkus.runtime.configuration.mappers.PropertyMapper.fromOption;
@@ -20,6 +25,16 @@ public class TransactionPropertyMappers implements PropertyMapperGrouping {
                 fromOption(TransactionOptions.TRANSACTION_XA_ENABLED_DATASOURCE)
                         .to("quarkus.datasource.\"<datasource>\".jdbc.transactions")
                         .transformer(TransactionPropertyMappers::getQuarkusTransactionsValue)
+                        .build(),
+                fromOption(TransactionOptions.TRANSACTION_DEFAULT_TIMEOUT)
+                        .to("quarkus.transaction-manager.default-transaction-timeout")
+                        .validator(value -> validateDuration(TransactionOptions.TRANSACTION_DEFAULT_TIMEOUT, value))
+                        .paramLabel("timeout")
+                        .build(),
+                fromOption(TransactionOptions.TRANSACTION_MIGRATION_TIMEOUT)
+                        .to("kc.spi-connections-jpa--quarkus--migration-transaction-timeout")
+                        .validator(value -> validateDuration(TransactionOptions.TRANSACTION_MIGRATION_TIMEOUT, value))
+                        .paramLabel("timeout")
                         .build()
         );
     }
@@ -32,5 +47,16 @@ public class TransactionPropertyMappers implements PropertyMapperGrouping {
         }
 
         return "enabled";
+    }
+
+    private static void validateDuration(Option<?> option, String value) {
+        try {
+            Duration duration = DurationConverter.parseDuration(value);
+            if (duration == null || duration.isNegative() || duration.isZero()) {
+                throw new PropertyException("Invalid duration '%s' for option '%s. Duration must be positive.".formatted(value, option.getKey()));
+            }
+        } catch (IllegalArgumentException e) {
+            throw new PropertyException("Invalid duration format '%s' for option '%s'. %s".formatted(value, option.getKey(), OptionsUtil.DURATION_DESCRIPTION));
+        }
     }
 }

--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/database/jpa/QuarkusJpaConnectionProviderFactory.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/database/jpa/QuarkusJpaConnectionProviderFactory.java
@@ -44,6 +44,7 @@ import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.KeycloakSessionFactory;
 import org.keycloak.models.dblock.DBLockManager;
 import org.keycloak.models.dblock.DBLockProvider;
+import org.keycloak.models.utils.KeycloakModelUtils;
 import org.keycloak.provider.ProviderConfigProperty;
 import org.keycloak.provider.ProviderConfigurationBuilder;
 import org.keycloak.provider.ServerInfoAwareProviderFactory;
@@ -53,6 +54,7 @@ import io.quarkus.arc.Arc;
 import io.quarkus.runtime.configuration.DurationConverter;
 import org.jboss.logging.Logger;
 
+import static org.keycloak.config.TransactionOptions.MIGRATION_TRANSACTION_TIMEOUT;
 import static org.keycloak.connections.jpa.util.JpaUtils.configureNamedQuery;
 import static org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction;
 import static org.keycloak.quarkus.runtime.storage.database.liquibase.QuarkusJpaUpdaterProvider.VERIFY_AND_RUN_MASTER_CHANGELOG;
@@ -66,6 +68,7 @@ public class QuarkusJpaConnectionProviderFactory extends AbstractJpaConnectionPr
     public static final String DEFAULT_PERSISTENCE_UNIT = "keycloak-default";
     private static final Logger logger = Logger.getLogger(QuarkusJpaConnectionProviderFactory.class);
     private static final String SQL_GET_LATEST_VERSION = "SELECT ID, VERSION FROM %sMIGRATION_MODEL ORDER BY UPDATE_TIME DESC";
+    private static final String MIGRATION_TRANSACTION_TIMEOUT_KEY = "migrationTransactionTimeout";
 
     enum MigrationStrategy {
         UPDATE, VALIDATE, MANUAL
@@ -106,6 +109,11 @@ public class QuarkusJpaConnectionProviderFactory extends AbstractJpaConnectionPr
         String schema = getSchema();
         boolean schemaChanged;
 
+        try {
+            KeycloakModelUtils.setTransactionLimit(factory, getMigrationTransactionTimeout());
+        } catch (Exception e) {
+            logErrorSettingMigrationTransactionTimeout(e);
+        }
         try (Connection connection = getConnection(); KeycloakSession session = factory.create()) {
             try {
                 try (Statement statement = connection.createStatement()) {
@@ -131,6 +139,13 @@ public class QuarkusJpaConnectionProviderFactory extends AbstractJpaConnectionPr
         } else {
             Version.RESOURCES_VERSION = id;
         }
+        // don't need to put this in a finally block as any exception thrown here will stop the server.
+        try {
+            // 0 means to revert to the default timeout.
+            KeycloakModelUtils.setTransactionLimit(factory, 0);
+        } catch (Exception e) {
+            logErrorSettingMigrationTransactionTimeout(e);
+        }
     }
 
     @Override
@@ -153,6 +168,12 @@ public class QuarkusJpaConnectionProviderFactory extends AbstractJpaConnectionPr
                 .name("migrationExport")
                 .type("string")
                 .helpText("Path for where to write manual database initialization/migration file.")
+                .add()
+                .property()
+                .name(MIGRATION_TRANSACTION_TIMEOUT_KEY)
+                .type("string")
+                .helpText("The transaction timeout for database migration transaction")
+                .defaultValue(MIGRATION_TRANSACTION_TIMEOUT)
                 .add()
                 .build();
     }
@@ -226,6 +247,7 @@ public class QuarkusJpaConnectionProviderFactory extends AbstractJpaConnectionPr
             operationalInfo.put("databaseUser", md.getUserName());
             operationalInfo.put("databaseProduct", md.getDatabaseProductName() + " " + md.getDatabaseProductVersion());
             operationalInfo.put("databaseDriver", md.getDriverName() + " " + md.getDriverVersion());
+            operationalInfo.put("migrationTimeout", getMigrationTransactionTimeout() + " seconds");
             logger.debugf("Database info: %s", operationalInfo.toString());
         } catch (SQLException e) {
             logger.warn("Unable to prepare operational info due database exception: " + e.getMessage());
@@ -349,5 +371,19 @@ public class QuarkusJpaConnectionProviderFactory extends AbstractJpaConnectionPr
         } catch (SQLException e) {
             logger.warnf(e, "Unable to validate %s 'isolation level' due to database exception", vendor);
         }
+    }
+
+    public int getMigrationTransactionTimeout() {
+        var value = config.get(MIGRATION_TRANSACTION_TIMEOUT_KEY, MIGRATION_TRANSACTION_TIMEOUT);
+        var duration =  DurationConverter.parseDuration(value);
+        // already validated by TransactionPropertyMappers
+        assert duration != null;
+        assert !duration.isZero();
+        assert !duration.isNegative();
+        return Math.toIntExact(duration.toSeconds());
+    }
+
+    private static void logErrorSettingMigrationTransactionTimeout(Exception e) {
+        logger.debug("Unable to set the transaction timeout for migration task. Using the default timeout.", e);
     }
 }

--- a/quarkus/runtime/src/main/resources/application.properties
+++ b/quarkus/runtime/src/main/resources/application.properties
@@ -12,9 +12,6 @@ quarkus.http.access-log.log-directory=${kc.home.dir:default}${file.separator}dat
 # Enables metrics from other extensions if metrics is enabled
 quarkus.datasource.metrics.enabled=${quarkus.micrometer.enabled:false}
 
-# Default transaction timeout
-quarkus.transaction-manager.default-transaction-timeout=300
-
 # Ignore split packages for Keycloak related packages
 quarkus.arc.ignored-split-packages=org.keycloak.*
 

--- a/quarkus/tests/integration/src/test-providers/java/org/keycloak/it/resource/realm/TestRealmResource.java
+++ b/quarkus/tests/integration/src/test-providers/java/org/keycloak/it/resource/realm/TestRealmResource.java
@@ -17,13 +17,19 @@
 
 package org.keycloak.it.resource.realm;
 
+import java.io.IOException;
+import java.util.Map;
+
+import com.arjuna.ats.arjuna.coordinator.TxControl;
 import org.infinispan.Cache;
 import org.infinispan.commons.configuration.io.ConfigurationWriter;
 import org.infinispan.commons.io.StringBuilderWriter;
 import org.infinispan.configuration.parsing.ParserRegistry;
 import org.jboss.logging.Logger;
 import org.keycloak.connections.infinispan.InfinispanConnectionProvider;
+import org.keycloak.connections.jpa.JpaConnectionProvider;
 import org.keycloak.models.KeycloakSession;
+import org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory;
 import org.keycloak.services.resource.RealmResourceProvider;
 
 import jakarta.ws.rs.GET;
@@ -32,6 +38,8 @@ import jakarta.ws.rs.PathParam;
 import jakarta.ws.rs.Produces;
 import jakarta.ws.rs.core.MediaType;
 import jakarta.ws.rs.core.Response;
+
+import org.keycloak.util.JsonSerialization;
 
 /**
  * @author Vaclav Muzikar <vmuzikar@redhat.com>
@@ -90,6 +98,21 @@ public class TestRealmResource implements RealmResourceProvider {
             new ParserRegistry().serialize(writer, cacheName, cache.getCacheConfiguration());
         }
         return Response.ok(out.toString(), MediaType.APPLICATION_JSON).build();
+    }
+
+    @Path("transaction/timeouts")
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response transactionTimeouts() throws IOException {
+        var jpaConnectionProvider = (QuarkusJpaConnectionProviderFactory) session.getKeycloakSessionFactory()
+                .getProviderFactory(JpaConnectionProvider.class);
+        var rsp = JsonSerialization.writeValueAsString(
+                Map.of(
+                        "default", TxControl.getDefaultTimeout(),
+                        "migration", jpaConnectionProvider.getMigrationTransactionTimeout()
+                )
+        );
+        return Response.ok(rsp, MediaType.APPLICATION_JSON_TYPE).build();
     }
 
     @Override

